Added the component specification (#604)
kubeflow/website: content/docs/pipelines/reference/component-spec.md

        preserved when modifying objects. See more information in the
        [Kubernetes user guide](http://kubernetes.io/docs/user-guide/annotations).

    * `labels`: A map of string keys and values that can be used to
        organize and categorize (scope and select) objects. May match selectors
        of replication controllers and services. See more information in the
        [Kubernetes user guide](http://kubernetes.io/docs/user-guide/labels).

### Interface

* `inputs` and `outputs`:
    Specifies the list of inputs/outputs and their properties. Each input or
    output has the following properties:

    * `name`: Human-readable name of the input/output. Name must be
        unique inside the inputs or outputs section, but an output may have the
        same name as an input.
    * `description`: Human-readable description of the input/output.
    * `default`: Specifies the default value for an input. Only
        valid for inputs.
    * `type`: Specifies the type of input/output. The types are used
        as hints for pipeline authors and can be used by the pipeline system/UI
        to validate arguments and connections between components. Basic types

>         are **String**, **Integer**, **Number**, **Boolean**.

Reviewer: Is this a full list of supported types? If not - can we have a reference here to the full list?

Author: Done.

            cases where the container always stores the output data in some
            hard-coded non-configurable local location. This property specifies
            a map between some outputs and local file paths where the program
            writes the output data files. Only needed for components that have
            hard-coded output paths. Such containers need to be fixed by
            modifying the program or adding a wrapper script that copies the
            output to a configurable location. Otherwise the component may be
            incompatible with future storage systems.

You can set all other Kubernetes container properties when you
use the component inside a pipeline.

## Using placeholders for command-line arguments

### Consuming input by value (parameter)

The placeholder is replaced by the contents of the input argument.

* In `component.yaml`: `{inputValue: Rounds}`
* Resulting Argo YAML: `"{{input.parameters.rounds}}"`

### Outputs

Output paths are filled in by the pipeline system. The Output construct is
replaced by a path. (The path can possibly point to a mounted output volume.)

* In `component.yaml': `{outputPath: Trained model}`

> * Resulting Argo YAML: `"/outputs/trained_model/data"`

Reviewer: is trained_model directory constructed from the "Trained model" name? In the docs - can we always keep the names matching? Or we need to reference a clear translation logic here.

Author: @Ark-kun

Reviewer 2: >is trained_model directory constructed from the "Trained model" name?

Yes.

> can we always keep the names matching?

We'll fix this part to make the names match. This is only a path though and the user does not interact with it.

> Or we need to reference a clear translation logic here.

We can describe the translation logic. But half of the purpose of generating the paths (as opposed to hard-coding the paths) is that it's opaque and can be changed later. We could as well just generate UUID paths like `/outputs/5c9b2d651e6c01cb41d834faa77925157d5d1599/data`. The user should not really care about or depend on some concrete logic.

Reviewer 2: Maybe we should not mention Argo here at all.

Reviewer 2: I've added comments with replacement text. We should not mention Argo. (Remove this line)

Author: Done.

            string or a placeholder. The most common placeholders are
            `{inputValue: Input name}` and `{outputPath: Output name}`.
        * `env`: Map of environment variables to set in the container.
        * `fileOutputs`: Legacy property that is only needed in
            cases where the container always stores the output data in some
            hard-coded non-configurable local location. This property specifies
            a map between some outputs and local file paths where the program
            writes the output data files. Only needed for components that have
            hard-coded output paths. Such containers need to be fixed by
            modifying the program or adding a wrapper script that copies the
            output to a configurable location. Otherwise the component may be
            incompatible with future storage systems.

You can set all other Kubernetes container properties when you
use the component inside a pipeline.

## Using placeholders for command-line arguments

### Consuming input by value (parameter)

The placeholder is replaced by the contents of the input argument.

> * In `component.yaml`: `{inputValue: Rounds}`

Reviewer 2: In `component.yaml`:
```command: [program.py, --rounds, {inputValue: Rounds}]```
In pipeline code:
```task1 = component1(rounds=150)```
Resulting command-line (the placeholder is substituted):
```program.py --rounds 150```

Author: Thanks. Done.

> * Resulting Argo YAML: `"{{input.parameters.rounds}}"`

Reviewer 2: ```suggestion
```

Author: Done.

### Outputs

> Output paths are filled in by the pipeline system. The Output construct is

Reviewer 2: ```suggestion
Output paths are filled in by the pipeline system. The outputPath placeholder is
```

Author: Done.

replaced by a path. (The path can possibly point to a mounted output volume.)

> 

Reviewer 2: ```suggestion
The parent directories of the path might or might not not exist.
The program must handle both cases without errors.

```

Author: Done.

> * In `component.yaml`: `{outputPath: Trained model}`

Reviewer 2: In `component.yaml`:
```command: [program.py, --out-model, {outputPath: trained_model}]```
In pipeline code:
```task1 = component1()```
Resulting command-line (the placeholder is substituted with generated path):
```program.py --out-model /outputs/trained_model/data```

Author: Done.
